etsi_its_conversion/etsi_its_conversion/src/Converter.cpp

```diff
@@ -462,22 +418,35 @@ void Converter::udpCallback(const UdpPacket::UniquePtr udp_msg) {
     else detected_etsi_type = "unknown";
   }
 
+  const uint8_t* protocol_version = reinterpret_cast<const uint8_t*>(&udp_msg->data[etsi_message_payload_offset_]);
+
   int msg_size = udp_msg->data.size() - etsi_message_payload_offset_;
-  ROS12_LOG(INFO, "Received ETSI message of type '%s' as bitstring (message size: %d | total payload size: %ld)", detected_etsi_type.c_str(), msg_size, udp_msg->data.size());
+  ROS12_LOG(INFO, "Received ETSI message of type '%s' (protocolVersion: %d) as bitstring (message size: %d | total payload size: %ld)", detected_etsi_type.c_str(), *protocol_version , msg_size, udp_msg->data.size());
 
   if (detected_etsi_type == "cam") {
 
-    // decode buffer to ROS msg
-    cam_msgs::CAM msg;
-    bool success = this->decodeBufferToRosMessage(&udp_msg->data[etsi_message_payload_offset_], msg_size, &asn_DEF_cam_CAM, std::function<void(const cam_CAM_t&, cam_msgs::CAM&)>(etsi_its_cam_conversion::toRos_CAM), msg);
-    if (!success) return;
-
-    // publish msg
+    if (*protocol_version == 2) { // CAM EN v1.4.1
+      cam_msgs::CAM msg;
+      bool success = this->decodeBufferToRosMessage(&udp_msg->data[etsi_message_payload_offset_], msg_size, &asn_DEF_cam_CAM, std::function<void(const cam_CAM_t&, cam_msgs::CAM&)>(etsi_its_cam_conversion::toRos_CAM), msg);
+      if (!success) return;
+#ifdef ROS1
+      publishers_["cam"]->publish(msg);
+#else
+      publisher_cam_->publish(msg);
+#endif
+    } else if (*protocol_version == 3) { // CAM TS v2.1.1
+      cam_ts_msgs::CAM msg;
+      bool success = this->decodeBufferToRosMessage(&udp_msg->data[etsi_message_payload_offset_], msg_size, &asn_DEF_cam_ts_CAM, std::function<void(const cam_ts_CAM_t&, cam_ts_msgs::CAM&)>(etsi_its_cam_ts_conversion::toRos_CAM), msg);
+      if (!success) return;
 #ifdef ROS1
-    publishers_["cam"]->publish(msg);
+      publishers_["cam_ts"]->publish(msg);
 #else
-    publisher_cam_->publish(msg);
+      publisher_cam_ts_->publish(msg);
 #endif
+    } else {
+      ROS12_LOG(ERROR, "Detected ETSI message type 'cam' with unknown protocol version %d, dropping message", *protocol_version);
+      return;
+    }
```

Comment:
@v0-e @joaomiguelpereiradealmeida

Do you happen to know how detection of CAM v1 vs CAM v2 is supposed to work? We thought to read it from the `protocolVersion` field of the `ItsPduHeader`, see this diff.
- [CAM EN v1.4.1](https://www.etsi.org/deliver/etsi_en/302600_302699/30263702/01.04.01_60/en_30263702v010401p.pdf) is specifying a value of `2`, see page 26
- [CAM EN v1.3.2](https://www.etsi.org/deliver/etsi_en/302600_302699/30263702/01.03.02_60/en_30263702v010302p.pdf) is specifying a value of `1`, see page 25
- [CAM TS v2.1.1](https://www.etsi.org/deliver/etsi_ts/103900_103999/103900/02.01.01_60/ts_103900v020101p.pdf) is not directly specifying a value
  - page 20, Section 7.2 directs to the CDD TS and annex B
  - annex B (page 23) directs to the [GitLab docs](https://forge.etsi.org/rep/ITS/asn1/cam_ts103900/-/blob/V2.1.1/docs/CAM-PDU-Descriptions.md)
  - there, also a `protocolVersion (2)` is mentioned

As you can see in this diff, right now we're switching to the CAM TS v2 if `protocolVersion == 3`.

Any ideas or suggestions?

FYI: @jpbusch

Reply:
The `protocolVersion` of 2 for both v1.4.1 and v2.1.1 is correct. It looks to me that you cannot distinguish between a v1.4.1 and v2.1.1 UPER binary message, as they essentially encode to the same thing. There may exist exceptions but I didn't found them yet, however I wasn't also exhaustive either.
Ultimately you will have to choose how you want to handle these messages for C-structs/ROS msgs reasons. Either default to v1.4.1 or v2.1.1 during conversion, perhabs with a configuration option.
The same goes for DENMs.